src/CVQualifiers.cpp

```diff
@@ -284,6 +284,28 @@ CVQualifiers::random_qualifiers(const Type* t, Effect::Access access,
 	return random_qualifiers(t, access, cg_context, no_volatile, RegularConstProb, RegularVolatileProb);
 }
 
+static bool is_volatile_ok_on_one_level(const Type* t)
+{
+	if (!CGOptions::lang_cpp()) return true;
+	bool volatile_ok = (t->eType != eStruct && t->eType != eUnion) || (t->has_assign_ops());
+	// Union with a struct field: we can't make it volatile, or we won't be able to assign to/from that field
+	if (volatile_ok && (t->eType == eUnion)){
+		for (size_t i = 0; i < t->fields.size(); ++i) {
+			const Type* field = t->fields[i];
+
+			if (field->eType == eStruct){
+				volatile_ok = false;
+				break;
+			}
+			else if (field->eType == eUnion){
+				volatile_ok = is_volatile_ok_on_one_level(field);
+				if (!volatile_ok) break;
+			}
+		}
+	}
+	return volatile_ok;
```

Comment:
Can we change the code to something like below?  In this way, we can reduce the level of nested ```if/for``` statements with all the early ```returns```.

    if (t->eType != eStruct && t->eType != eUnion) 
        return true; 
    // Union with a struct field: we can't make it volatile, or we won't be able to assign to/from that field
    if (!t->has_assign_ops()) 
        return false;

     if (t->eType != eUnion)
        return true;

    for (size_t i = 0; i < t->fields.size(); ++i) {
        const Type* field = t->fields[i];
        if (field->eType == eStruct) 
            return false;
        if (field->eType == eUnion && !is_volatile_ok_on_one_level(field))
            return false;
    }   

    return true;

Hopefully my interpretation to the original code logic would be correct.